### ISSUE TITLE
BDD implementation with Playwright test runner

### ISSUE THREAD

**Issue by ShanmukhaGajula** [microsoft/playwright#11975]
Our team has a case to use the BDD test, Playwright has any plan to add their own BDD implementation using the playwright test runner? That would be great if to have.

This can help many people who want to use BDD along with Playwright. That's where cucumber became popular 😉

**Comment by unlikelyzero** [comment]
@1996phani would https://codecept.io/playwright/ serve your needs?

**Comment by saranyakakarlapudi** [comment]
BDD implementation with Playwright test runner is what am also looking for..

**Comment by keshavmunagala** [comment]
It would be great to have BDD with Playwright

**Comment by 1996phani** [comment]
> @1996phani would https://codecept.io/playwright/ serve your needs?

This package has been deprecated

I am looking for something kind of Native support from the Playwright

**Comment by pavelfeldman** [comment]
Let's collect upvotes. Please upvote the OP!

**Comment by CodeStrikerr** [comment]
Yeah BBD is a must thing!

**Comment by soundaryakk** [comment]
It would be helpful ,if we have BDD along with Playwright.

**Comment by kvsrk432** [comment]
BDD support is a perk

**Comment by siddharththakurji** [comment]
bdd is a good feature to have.. i am using in selenium java.. i need it in playwright to migrate

**Comment by Vinaymanchala** [comment]
I'm looking for bdd too!

**Comment by Geetha2604** [comment]
Good to have BDD with playwright

**Comment by dimmetasowjanya** [comment]
Having BDD with Playwright will be helpful

**Comment by unlikelyzero** [comment]
> > @1996phani would https://codecept.io/playwright/ serve your needs?
> 
> @unlikelyzero,  The package you're mentioning has been deprecated.
> 
> I am looking for something kind of Native support from the Playwright

@ShanmukhaGajula where does it say that?

**Comment by ShanmukhaGajula** [comment]
> > > @1996phani would https://codecept.io/playwright/ serve your needs?
> > 
> > 
> > @unlikelyzero,  The package you're mentioning has been deprecated.
> > I am looking for something kind of Native support from the Playwright
> 
> @ShanmukhaGajula where does it say that?

I see a pop-up yesterday when I opened that link. I think they've removed now.  However, I see playwright test runner and reporting is very good and hence requesting Playwright team for that feature within playwright itself.

**Comment by abhirammocharla** [comment]
BDD is always good to have

**Comment by PranayEtikala** [comment]
Yes, Looking for BDD to implement playwright in our project.

**Comment by Akhila4062** [comment]
BDD is a must-have feature for us too! It really helps.

**Comment by Mohan2736** [comment]
Hope playwright will add this BDD feature..

**Comment by ShanmukhaGajula** [comment]
> Let's collect upvotes. Please upvote the OP!

@pavelfeldman, I think we have a lot of demand for it. as more requests coming up for BDD 😃. Hope you will implement this similar to Cucumber or better than it.

**Comment by marko-simic** [comment]
I'm looking for BDD too!

**Comment by IsabelFrancis** [comment]
Looking forward to have BDD feature with Playwright,

**Comment by KalyanPuppala36** [comment]
It would be great to have BDD with Playwright

**Comment by KalyanPuppalaSRM** [comment]
It would be helpful ,if we have BDD along with Playwright.

**Comment by SankarPuppala** [comment]
It would be nice to have BDD along with Playwright.

**Comment by LoneyR001** [comment]
BDD with Playwright! That sounds great. Looking forward to it.

**Comment by macroking** [comment]
@pavelfeldman  Looks like 90% of the comments and upvotes are well planned, just digging through the user profile, most of the users just opened GitHub account to upvote this request.

May be one particular group of users really need this!

**Comment by ShanmukhaGajula** [comment]
Hey @macroking, I've started following this thread yesterday by looking at the request of BDD. I don't who the hell are others. I've started with Cypress and looking for Playwright to be mature to use it. Do you think all Playwright users have Github accounts what if by looking at this thread and starting commenting on it by signing up? Stop being a sherlock holmes, when i go through the comments I can see a guy who suggested codecept and your profile has codecept repo too

**Comment by Nightfall77** [comment]
> @pavelfeldman Looks like 90% of the comments and upvotes are well planned, just digging through the user profile, most of the users just opened GitHub account to upvote this request.
> 
> May be one particular group of users really need this!

Hi Just created account to comment in this thread, Hope this works !😜

**Comment by Nightfall77** [comment]
> @pavelfeldman Looks like 90% of the comments and upvotes are well planned, just digging through the user profile, most of the users just opened GitHub account to upvote this request.
> 
> May be one particular group of users really need this!

Hi even I need BDD feature.. I just created account to be able to comment to this thread. Hope it is fine with you 😜
P.S: I don't know who others are!

**Comment by kumars3** [comment]
Playwright has already got support to write BDD tests using cucumber.js(dependency =  "@cucumber/cucumber": "^7.3.1",). I am just wondering whether **PlaywrightTestConfig** (dependency = '@playwright/test') is supported in BDD tests(GIVEN, WHEN, THEN because the global configuration (const config: PlaywrightTestConfig = {
  use: {.......) is not being picked automatically as written in https://playwright.dev/docs/test-configuration#global-configuration eventhough I am using playwright test runner.

Looks like the **PlaywrightTestConfig** works only with 'test' block in .ts or .js file. Can somebody confirm this how to call **PlaywrightTestConfig** in BDD tests using playwright test runner.

**Comment by isherkhan** [comment]
Cucumber integration with playwright. Looking forward for it.

**Comment by nbnds** [comment]
Definitely a needed feature.

**Comment by NikkTod** [comment]
Cucumber integration with playwright. +1

**Comment by kumars3** [comment]
An official cucumber Playwright adapter is needed to use all the features of playwright which works on 'test' block i.e. with @playwright/test test runnier.
Please see the comment of @mxschmitt:

https://github.com/microsoft/playwright/issues/12069#issuecomment-1047112096

**Comment by jgrieger** [comment]
Definitely needed. Integrating with e.g. Cucumber as the test runner is hacky.

**Comment by testgitdl** [comment]
Do you know when the cucumber integration will be done? This is really needed as not being able to run the tests with playwright test runner is really limiting

**Comment by mathew-jithinm** [comment]
Would be really great if we have an official support for BDD, for syntaxes like Cucumber, etc.

**Comment by pixero** [comment]
We are also very much waiting for bdd support in playwright. 
Now we use cucumber runner and playwright. But I really wanted to be able to also use plywright.config in this case

**Comment by lsargent0** [comment]
+1 for adding Playwright Test Cucumber integration

**Comment by asdfsafd** [comment]
Pls consider the Playwright/Test-Cucumber integration

**Comment by PavanMudigondaTR** [comment]
It would be awesome to use Playwright Test with Cucumber BDD.

**Comment by natalygoloborodko** [comment]
Adding votes on behalf of my company for this feature.

**Comment by uzeyirqa** [comment]
Are there any plans to implement this feature request?

**Comment by benjefferies** [comment]
@kumars3 is perfectly right with https://github.com/microsoft/playwright/issues/11975#issuecomment-1038297911. It's not completely necessary to implement a BDD framework for Playwright however hooks into Playwrights' initialisation need to be exposed so that the Playwright context can be initialised. I've done some digging to try and figure out a way and came up with nothing. 

Does anyone know of a workaround here so in setups within other test runners' `@playwright/test` context can be initialised?

**Comment by uzeyirqa** [comment]
> Does

I initialize the browser and context in a beforeAll hook with cucumber js. As for settings like timeout  they get picked up from playwright.config.js the only issue I’m having is that I need to configure video and trace recording manually in the hooks they are not picked up from the playwright config.

**Comment by uzeyirqa** [comment]
> @kumars3 is perfectly right with [#11975 (comment)](https://github.com/microsoft/playwright/issues/11975#issuecomment-1038297911). It's not completely necessary to implement a BDD framework for Playwright however hooks into Playwrights' initialisation need to be exposed so that the Playwright context can be initialised. I've done some digging to try and figure out a way and came up with nothing.
> 
> Does anyone know of a workaround here so in setups within other test runners' `@playwright/test` context can be initialised?

I initialize the browser and context in a beforeAll hook with cucumber js. As for settings like timeout they get picked up from playwright.config.js the only issue I’m having is that I need to configure video and trace recording manually in the hooks they are not picked up from the playwright config.

**Comment by benjefferies** [comment]
> > @kumars3 is perfectly right with [#11975 (comment)](https://github.com/microsoft/playwright/issues/11975#issuecomment-1038297911). It's not completely necessary to implement a BDD framework for Playwright however hooks into Playwrights' initialisation need to be exposed so that the Playwright context can be initialised. I've done some digging to try and figure out a way and came up with nothing.
> > Does anyone know of a workaround here so in setups within other test runners' `@playwright/test` context can be initialised?
> 
> I initialize the browser and context in a beforeAll hook with cucumber js. As for settings like timeout they get picked up from playwright.config.js the only issue I’m having is that I need to configure video and trace recording manually in the hooks they are not picked up from the playwright config.

I'm experiencing the same issue. It also affects the `toMatchSnapshot` assertion.

@pavelfeldman this is the 3rd most popular issue on the playwright. Is there any workaround you know of to use another test runner by initialising `@playwright/test` context in a similar manner?

**Comment by mytechthingz** [comment]
A lot of companies won't consider using Playwright Test without Cucumber because they would be migrating from their current framework which already supports Cucumber and BDD, and it's a part of their development process works so it's a necessity

**Comment by ForQA02** [comment]
@pavelfeldman  is it possible to get an update on this please? Even if that's to give a rough timeline or just to say it's not currently on the horizon. This will help people, such as myself, who are looking at automation tools to decide whether they can initially work without Cucumber and integrate it later knowing it's coming within a certain timeframe, or that they just need to use a different tool.

**Comment by PavanMudigonda** [comment]
> 

I am also in the same situation, not able to ask teams to use Playwright as on another side I was promoting people to use Cucumber BDD frameworks ! A combination of both would be awesome! I am certainly not interested to use Tallyb's framework as it's becoming difficult to use it with the latest playwright features! But she certainly did good job making one for people!

**Comment by PavanMudigondaTR** [comment]
@pavelfeldman @mxschmitt  Please let us know if this will be taken up anytime soon?

**Comment by holdfenytolvaj** [comment]
Is there testing without BDD?

**Comment by gzuppl** [comment]
I'm in the same boat as many of the earlier posters: We would migrate to playwright instantly if there was "native support" for BDD - as in an adapter for Cucumber resulting in all playwright features being usable in tandem with test cases written in Gherkin.

**Comment by PavanMudigondaTR** [comment]
I have ended up recommending to use WebdriverIO for one of the team who is currently using proctractor cucumber framework. It would have been awesome if playwright test can offer cucumber out of box. May be in future I can recommend some teams to use Playwright Test instead.

**Comment by dsumac** [comment]
In the mean time, it exists an acceptable alternative: use **the playwright `steps`** to explain all Given/When/Then assertions.

The solution is light and fonctionnal.

Ex:

```
import { test, expect } from '@playwright/test';

test('Scenario explanation', async ({ page, context  }) => {

  await test.step('Given ...', async () => {
    ...
  });

  await test.step('When user make a stuff', async () => {
    ...
  });

  await test.step('Then a stuff is displayed on page', async () => {
    ...
  });

  await test.step('And user click on a button', async () => {
    ...
  });

  await test.step('Then an another stuff is displayed', async () => {
    ...
  });
});

```

**Comment by PavanMudigondaTR** [comment]
> In the mean time, it exists an acceptable alternative: use **the playwright `steps`** to explain all Given/When/Then assertions.
> 
> 
> 
> The solution is light and fonctionnal.
> 
> 
> 
> Ex:
> 
> 
> 
> ```
> 
> import { test, expect } from '@playwright/test';
> 
> 
> 
> test('Scenario explanation', async ({ page, context  }) => {
> 
> 
> 
>   await test.step('Given ...', async () => {
> 
>     ...
> 
>   });
> 
> 
> 
>   await test.step('When user make a stuff', async () => {
> 
>     ...
> 
>   });
> 
> 
> 
>   await test.step('Then a stuff is displayed on page', async () => {
> 
>     ...
> 
>   });
> 
> 
> 
>   await test.step('And user click on a button', async () => {
> 
>     ...
> 
>   });
> 
> 
> 
>   await test.step('Then an another stuff is displayed', async () => {
> 
>     ...
> 
>   });
> 
> });
> 
> 
> 
> ```
> 
> 
> 
> And the report (with html reporter) is very readable. 


Yes, I am using this currently. Currently this approach is causing me to repeat code and maintenance difficult. Just trying to see if cucumber support would remove repeating same code related to given when then. Any how thanks for inputs.

**Comment by TakesTheBiscuit** [comment]
Up

**Comment by tonirecio** [comment]
Up

**Comment by dsumac** [comment]
> > In the mean time, it exists an acceptable alternative: use **the playwright `steps`** to explain all Given/When/Then assertions.
> > The solution is light and fonctionnal.
> > Ex:
> > ```
> > 
> > import { test, expect } from '@playwright/test';
> > 
> > 
> > 
> > test('Scenario explanation', async ({ page, context  }) => {
> > 
> > 
> > 
> >   await test.step('Given ...', async () => {
> > 
> >     ...
> > 
> >   });
> > 
> > 
> > 
> >   await test.step('When user make a stuff', async () => {
> > 
> >     ...
> > 
> >   });
> > 
> > 
> > 
> >   await test.step('Then a stuff is displayed on page', async () => {
> > 
> >     ...
> > 
> >   });
> > 
> > 
> > 
> >   await test.step('And user click on a button', async () => {
> > 
> >     ...
> > 
> >   });
> > 
> > 
> > 
> >   await test.step('Then an another stuff is displayed', async () => {
> > 
> >     ...
> > 
> >   });
> > 
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And the report (with html reporter) is very readable.
> 
> Yes, I am using this currently. Currently this approach is causing me to repeat code and maintenance difficult. Just trying to see if cucumber support would remove repeating same code related to given when then. Any how thanks for inputs.

To avoid code repeating on "given/when/then" you can extract the steps like this:

```
async function checkStep(page: Page) {
  await test.step('When ... common step', async () => {
    await ...
  });
}
```
and use it in tests:

```
test('My test which is extracted steps', async ({ page }) => {
  await checkStep(page);
});
```

**Comment by natalygoloborodko** [comment]
Using test.step() doesn't solve the problem.
Many companies use protractor + cucumber and business wants to see those Gherkins as they were after switching to Playwright. 
It may work for completely new projects, but for migrating existing ones... no.

**Comment by PavanMudigondaTR** [comment]
> Using test.step() doesn't solve the problem.
> 
> Many companies use protractor + cucumber and business wants to see those Gherkins as they were after switching to Playwright. 
> 
> It may work for completely new projects, but for migrating existing ones... no.

Right now I need for two business units who are on proctractor and selenium with cucumber. So test.step is not solving it. I am trying WebdriverIO but it's not as good as playwright but I am not having choice other than going with WebdriverIO .

**Comment by mahula** [comment]
Up

**Comment by revathyvarma12** [comment]
is there any way to run cucumber test parallel with playwright

**Comment by benjefferies** [comment]
> is there any way to run cucumber test parallel with playwright

Cucumber takes an argument `--parallel 3` which you can use to run tests in parallel https://github.com/cucumber/cucumber-js/blob/main/docs/parallel.md

**Comment by oliverxu2008** [comment]
yes, bdd is good to have for playwright.

**Comment by mortelee** [comment]
The strong point of using cucumber is that you use feature files to explain what is tested. Using test.step and even extracting steps so that they can be re-used is a small step for software engineers to make the test better readable but is still not that what we find in cucumber and using feature files.
Tally's example of how to use cucumber icw playwright works for me (https://github.com/Tallyb/cucumber-playwright) and I don't understand why so many people want to have native support from playwright/test

**Comment by lthurr** [comment]
The suggestion from @mortelee works fine. 👍 

However, there are some things that we can't use **outside** of Playwright runner, like:
```
.toMatchSnapshot(name[, options])
.toMatchSnapshot([options])
.toHaveScreenshot(name[, options])
.toHaveScreenshot([options])
```

Visual comparison is a really nice functionality of playwright that we can't use if we happen to have a different runner on our projects.

**Comment by benjefferies** [comment]
> The suggestion from @mortelee works fine. 👍
> 
> However, there are some things that we can't use **outside** of Playwright runner, like:
> 
> ```
> .toMatchSnapshot(name[, options])
> .toMatchSnapshot([options])
> .toHaveScreenshot(name[, options])
> .toHaveScreenshot([options])
> ```
> 
> Visual comparison is a really nice functionality of playwright that we can't use if we happen to have a different runner on our projects.

To get around the snapshot stuff I created a function

```
function toMatchSnapshot(name: string, actual: any) {
  const file = `${process.cwd()}/snapshots/${name}.json`;
  if (!fs.existsSync(file)) {
    fs.writeFileSync(file, JSON.stringify(actual, null, 2));
  }
  const expected = require(file);
  expect(actual).toMatchObject(expected);
}
```

**Comment by yaffol** [comment]
I am using https://github.com/Tallyb/cucumber-playwright to integrate Playwright and Cucumber. The biggest thing for me that is missing is the integration with the VSCode extension - I liked being able to run/debug tests direct from the test explorer in VSCode, whereas now I must run and debug all my tests via the command line. This isn't a huge difficulty, but it's a shame, and it does make for more friction when working with tests. I'd like to be able to run my BDD/gherkin-based tests from within test explorer as well.

**Comment by mechandan** [comment]
Yes this feature is must , microsoft team please bring a native support playwright + BDD … thanks

**Comment by evertonlperes** [comment]
I agree with the folks, this feature will help a lot of people where BDD is a requirement.

Currently I'm in the same situation as many folks here, using cucumber-js with a customized wold config (a bit overcomplicated due multiple browsers vs devices).

Hope some solution comes up soon. 
Excited to be a beta tester for it. I'm currently replacing an old Java8+Selenium+Cucumber framework with Typescript + Playwright + Cucumber (around 5K test scenarios).

**Comment by JoaoEnesGoncalves** [comment]
Please, please! 
playwright with a native integration with cucumber/BDD/Gherkin it will be extremely helpful !

**Comment by NikkTod** [comment]
Hello great PW team, this has over 150 votes, probably the most voted thing , how much more are needed  ;)  

@mxschmitt @aslushnikov

**Comment by SpiderRam** [comment]
Adding another vote for this.  Strongly second the comment about losing the functionality of the extensions.  And it seems that the custom error messaging on expects does not come through, either.  If I throw an error, it shows up as expected.  But the [custom expect message](https://playwright.dev/docs/test-assertions#custom-expect-message) is lost completely.

**Comment by PavanMudigondaTR** [comment]
@mxschmitt @aslushnikov

Please let us know your opinion on this feature request ? I was wondering if this will happen in next 12 months !

**Comment by Tallyb** [comment]
> I am using https://github.com/Tallyb/cucumber-playwright to integrate Playwright and Cucumber. The biggest thing for me that is missing is the integration with the VSCode extension - I liked being able to run/debug tests direct from the test explorer in VSCode, whereas now I must run and debug all my tests via the command line. This isn't a huge difficulty, but it's a shame, and it does make for more friction when working with tests. I'd like to be able to run my BDD/gherkin-based tests from within test explorer as well.

@yaffol My demo repo contains a way to run the tests directly from vscode, by using the debugger: https://github.com/Tallyb/cucumber-playwright#in-visual-studio-code
(for the record: I also upvoted this feature, but until we have it, you are welcome to use my repo).

**Comment by csabasolyom** [comment]
> > The suggestion from @mortelee works fine. 👍
> > However, there are some things that we can't use **outside** of Playwright runner, like:
> > ```
> > .toMatchSnapshot(name[, options])
> > .toMatchSnapshot([options])
> > .toHaveScreenshot(name[, options])
> > .toHaveScreenshot([options])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Visual comparison is a really nice functionality of playwright that we can't use if we happen to have a different runner on our projects.
> 
> To get around the snapshot stuff I created a function
> 
> ```
> function toMatchSnapshot(name: string, actual: any) {
>   const file = `${process.cwd()}/snapshots/${name}.json`;
>   if (!fs.existsSync(file)) {
>     fs.writeFileSync(file, JSON.stringify(actual, null, 2));
>   }
>   const expected = require(file);
>   expect(actual).toMatchObject(expected);
> }
> ```
> 
> ```
> Then('the BEE configuration has rules', function (dataTable: DataTable) {
>   const [{ id = 1 }] = dataTable?.hashes()?.length === 0 ? [{}] : dataTable.hashes();
>   const name = id ? `${this.cucumberScenarioName}-${id}-rules` : `${this.cucumberScenarioName}-rules`;
>   toMatchSnapshot(name, this.beeConfiguration.rules);
> });
> ```

@benjefferies Have you been able to use this for image comparison? 
Could you give some more context on this? e.g. what is in beeConfiguration.rules and why are you loading .json file?
thanks

**Comment by benjefferies** [comment]
@csabasolyom I haven't use it for image comparison. All I've done is create a custom snapshot assertion check similar to https://jestjs.io/docs/snapshot-testing

**Comment by refactoreric** [comment]
Built-in support for BDD would be great. Pure frontend tests can do without, but for real end-to-end tests on more complex applications, you need more layers in your test automation architecture. Something like specs/feature files (tests), step definitions, page objects and API clients. It doesn't necessarily have to be Cucumber

**Comment by josempd** [comment]
This would be great to have! I'm also using another implementation, but BDD is important for scalability.

**Comment by AbhishekPalani** [comment]
I keep coming back to this issue thread everyday hoping for a confirmation/ETA/resolution from the Playwright Team. A native support for BDD style playwright test runner will encourage many QAs to migrate existing selenium or Cypress based automation solutions.

**Comment by Agustin-DeLuca** [comment]
> I keep coming back to this issue thread everyday hoping for a confirmation/ETA/resolution from the Playwright Team. A native support for BDD style playwright test runner will encourage many QAs to migrate existing selenium or Cypress based automation solutions. 

I love Playwright but I really can't understand why the devs won't say anything about their most commented issue.

**Comment by GeGerardo** [comment]
Hi @pavelfeldman how are you doing?
Do you have any update regarding this feature request?

**Comment by TimvdEijnden** [comment]
When comparing webdriverIO & cypress support cucumber through the packages:
@badeball/cypress-cucumber-preprocessor
@wdio/cucumber-framework

Using cucumber-js it is possible but you'll miss all the advantages of the playwright test runner.
So an implementation within playwright increase the change ofus migrating to it :)

**Comment by PavanMudigondaTR** [comment]
> > I keep coming back to this issue thread everyday hoping for a confirmation/ETA/resolution from the Playwright Team. A native support for BDD style playwright test runner will encourage many QAs to migrate existing selenium or Cypress based automation solutions.
> 
> I love Playwright but I really can't understand why the devs won't say anything about their most commented issue.

This could be complex and complicated to implement otherwise they would done it already as its been a year now since this request was originally logged.

**Comment by Selenium-ToolBox** [comment]
I really love playwright, but sadly because of this feature not being implemented, I can't offer this as a Test Automation solution for my company.

**Comment by arvinder06** [comment]
Please add this feature.

**Comment by MartinB134** [comment]
The handling of the page object from the session is far from perfect and also adding new URLs to the pytest objects as property should be changed (don't know how to do it better), but this is the only option for now, if you want to have BDD, Data Driven testing, Playwright and Allure reporting. I know there is much work to do (dead code, IDE integration). But my resources are limited during my free time:
https://github.com/MartinB134/playwright-pytest
I will appreciate any Feedback and participation!

**Comment by 3ackdoor** [comment]
Do you guys @pavelfeldman @aslushnikov have any update for this request? This has been requested a year for now so.

**Comment by avinashdj** [comment]
Any udpate on this.  Eagerly waiting.

**Comment by VitaliyPotapov** [comment]
While waiting official comment I'd suggest to discuss, what actually do you expect here. I see several options:

1. Using playwright test-runner that reads and executes BDD tests written in gherkin
2. Using cucumber test-runner that have easy way to integrate playwright in steps

**Comment by mechandan** [comment]
> 1. Using

Yes we can do that as well. I would prefer playwright test runner if we can make it in anyways 😊

**Comment by bluvulture** [comment]
Hello, 

Ideally this would be also backwards compatible with current tests, meaning steps or assertions could be marked as a BDD step. 

Use case:
We have already added cucumber-js within our test suite cause it was a business requirement, and we are using same test base with cucumber and playwright runner. 
Cucumber is used for business people, playwright for everything else. 
I would hate to need to rewrite everything that we have currently written to fit it in, would probably stick to current implementation in that case.

**Example** 

Base function
```
export const clickAgreeWithDisclaimer = async ({ page }: TestArgs): Promise<void> => {
    await page.getByTestId(AGREE_BUTTON).click();
};
```

Cucumber feature file 

```
  Background:
    Given I agree with disclaimer
```

BDD step

`Given('I agree with disclaimer', () => agreeWithDisclaimer({ page, context }));`

Playwright

```
test.describe('Feature: Overview', () => {
    test.beforeEach(async ({ page, context }) => {
        await agreeWithDisclaimer({ page, context });
    });
});
```

**Comment by yaffol** [comment]
If I was to describe my ideal developer experience in terms of using the tools, it would be to be able to use playwright's vs-code extension to see and run my tests, with the great tooling like tracing, but to be able to run tests derived from BDD features there.

And also to be able to run them via the CLI, but that goes without saying.

I guess what I am trying to say is that I really appreciate the DX (Testing eXperience?) of playwright/test, the trace viewer and now UI ('watch') mode, and would really like to be able to use them **as well as ** tests that are executable specifications.

**Comment by NikkTod** [comment]
Hey guys @dgluzman, @pavelfeldman, @aslushnikov, have you ever discussed that feature in your team meetings, please please share some update on it, many people are wondering if there is a chance to see it.

**Comment by vitalets** [comment]
I've implemented BDD testing with Playwright runner for Node.js. Published it as [playwright-bdd](https://github.com/vitalets/playwright-bdd) npm package.
It provides `bddgen` CLI that allows to convert existing Cucumber features into Playwright tests:
<details>
<summary>Example of generated test</summary>

  From
  ```gherkin
  Feature: Playwright site

       Scenario: Check title
           Given I open url "https://playwright.dev"
           When I click link "Get started"
           Then I see in title "Playwright"
  ```

  To
  ```js
  import { test } from 'playwright-bdd';

  test.describe('Playwright site', () => {

    test('Check title', async ({ Given, When, Then }) => {
      await Given('I open url "https://playwright.dev"');
      await When('I click link "Get started"');
      await Then('I see in title "Playwright"');
    });

  });    
  ```
</details>

When tests are generated Playwright config can read them and run as usual. For each test `playwright-bdd` creates Playwright-powered Cucumber World and passes it to step definitions. This allows to use Playwright objects (e.g. `page`) in Cucumber steps:

<details>
<summary>Example of step definition</summary>

  ```ts
  import { expect } from '@playwright/test';
  import { Given, When, Then } from '@cucumber/cucumber';
  import { World } from 'playwright-bdd';

  Given('I open url {string}', async function (this: World, url: string) {
    await this.page.goto(url);
  });

  When('I click link {string}', async function (this: World, name: string) {
    await this.page.getByRole('link', { name }).click();
  });

  Then('I see in title {string}', async function (this: World, keyword: string) {
    await expect(this.page).toHaveTitle(new RegExp(keyword));
  });  
  ```
</details>

Finally, you can run BDD tests as a single command:
```
npx bddgen && npx playwright test
```
I'd like if someone try and share feedback on that.

**Comment by PavanMudigondaTR** [comment]
> 

@vitalets I am going to try this today and I will let you know by Monday.

**Comment by PavanMudigondaTR** [comment]
> 

@vitalets 
It seems okay to me at high level. I will spend more time in coming day to explore If I could use this.  But could you tell me if we can use the Outline Feature with test data in feature files ?

**Comment by vitalets** [comment]
> @vitalets It seems okay to me at high level. I will spend more time in coming day to explore If I could use this. But could you tell me if we can use the Outline Feature with test data in feature files ?

Yes. Scenario Outline is [supported](https://github.com/vitalets/playwright-bdd/blob/main/test/gen/scenario-outline/sample.feature) and converts into several tests with the same name.

**Comment by PavanMudigonda** [comment]
> > @vitalets It seems okay to me at high level. I will spend more time in coming day to explore If I could use this. But could you tell me if we can use the Outline Feature with test data in feature files ?
> 
> 
> 
> Yes. Scenario Outline is [supported](https://github.com/vitalets/playwright-bdd/blob/main/test/gen/scenario-outline/sample.feature) and converts into several tests with the same name.

This is awesome achievement. I wish playwright team can accept your solution as community contribution. I am going to test this in detail next week.

**Comment by PavanMudigondaTR** [comment]
> 

@vitalets 
I realized the tests are getting duplicated when I used this pattern that the only caveat I have seen so far. I use an automation pattern to synchronize tests from Feature Files to Azure DevOps TestPlans that stores the parameters and does't duplicate.

Since it will another year or two  for playwright team to implement this feature request officially, I would like to start using your pattern. I hope any migration to new framework would be easier later on.

**Comment by vitalets** [comment]
> @vitalets I realized the tests are getting duplicated when I used this pattern that the only caveat I have seen so far. I use an automation pattern to synchronize tests from Feature Files to Azure DevOps TestPlans that stores the parameters and does't duplicate.

@PavanMudigondaTR could you create an issue with details and we will think what we can do. Maybe wrap into another `test.describe`..

**Comment by nguyentramy87** [comment]
Any update for this feature, please?

**Comment by gegoncalves** [comment]
@pavelfeldman @aslushnikov
Sorry to disturb you guys. Does PW team has any plan to include this one in the roadmap?

Thanks for the good work 😄

**Comment by Niitch** [comment]
I've created a wrapper that makes the Playwright test runner handle the Gherkin syntax.
It's published here : [gherkin-wrapper](https://github.com/Niitch/gherkin-wrapper)

The package allows to register "step definition" functions in a Cucumber-like fashion and then run your tests using the `playwright test` CLI.

**Comment by jan-molak** [comment]
Hello all! It's great to see so many people interested in BDD and making Playwright support Gherkin syntax! However, I'm not sure if making Playwright Test dynamically generate code is even necessary.

Allow me to elaborate. Cucumber is _much_ more than just a Gherkin parser; It has a runner that offers [parallel test execution](https://github.com/cucumber/cucumber-js/blob/main/docs/parallel.md), supports tagging and filtering scenarios using [tag expressions](https://github.com/cucumber/tag-expressions), integrates with [numerous reporters](https://cucumber.io/docs/cucumber/reporting/?lang=javascript#custom-formatters), third-party [plugins](https://cucumber.io/docs/cucumber/reporting/?lang=javascript#third-party-plugins), not to mention project management tools, [code editors](https://behavepro.app/features) and [collaboration tools](https://smartbear.com/blog/introducing-cucumberstudio/). Cucumber.js also has a dedicated [VS Code extension](https://marketplace.visualstudio.com/items?itemName=CucumberOpen.cucumber-official) that supports Gherkin auto-completion.

Making Playwright Test try to replicate all that seems like a _lot_ of work that could've been avoided by simply using Cucumber.js with Playwright and playing to the strengths of both tools. In fact, this is what I do with [Serenity/JS](https://serenity-js.org) and what I'll attempt to show in this post.

**TL;DR** - check out https://github.com/serenity-js/serenity-js-cucumber-playwright-template

## Integrating Cucumber.js with Playwright 

What I love about Playwright is that we can integrate it with Cucumber.js just like any other library. 

To do that, we need to:
1. launch the Playwright browser before we start the test run (we only need one browser since we can achieve test isolation using [Playwright Contexts](https://playwright.dev/docs/browser-contexts#how-playwright-achieves-test-isolation)),
2. create a Context and a Page for each scenario and make them available in Cucumber step definitions,
3. close the Playwright browser when the test run is finished.

We can achieve 1. and 3. using `BeforeAll` and `AfterAll` hooks:

```typescript
import {AfterAll, BeforeAll} from '@cucumber/cucumber';
import * as playwright from 'playwright';

let browser: playwright.Browser;

BeforeAll(async () => {
    browser = await playwright.chromium.launch({
        headless: true,
    });
});

AfterAll(async () => {
    if (browser) {
        await browser.close();
    }
});
```

If we wanted to make it more fancy, we could use a custom [World](https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/world.md#real-world-example) and use lazy initialisation to launch the browser and initialise the `page`:

```typescript
import {AfterAll, Before, setWorldConstructor, World} from '@cucumber/cucumber';
import * as playwright from 'playwright';

class CustomWorld extends World {
    private static browser?: playwright.Browser;
    private page: playwright.Page;

    async getPage(): Promise<playwright.Page> {
        if (! this.page) {
            const browser = await this.getBrowser();
            const context = await browser.newContext();
            this.page = await context.newPage();    
        }
        return this.page;
    }

    private async getBrowser(): Promise<playwright.Browser> {
        if (! CustomWorld.browser) {
            CustomWorld.browser = await playwright.chromium.launch({
                headless: true,
            });
        }

        return CustomWorld.browser;
    }

    static async shutDown(): Promise<void> {
        await CustomWorld.browser?.close()
    }
}

setWorldConstructor(CustomWorld);

AfterAll(async () => {
    await CustomWorld.shutDown();
})
```

With the above in place, we can define Cucumber steps like this:

```typescript
Given('I want to learn about Playwright', async function (this: CustomWorld) {
    const page = await this.getPage()
    await page.goto('https://playwright.dev')
    // and so on
})
```

## Limitations

Since the [Playwright extension for VS Code](https://playwright.dev/docs/getting-started-vscode) relies on Playwright Test, it doesn't support running tests with standard Cucumber.js.

What would be awesome 🤩 is if developers of other VS Code extensions, such as Cucumber.js, could somehow plug into the excellent toolchain offered by the Playwright extension for VS Code to run and debug test scenarios. 

I'm not sure how we could do it, but I'd be happy to contribute to this effort.

**Comment by yaffol** [comment]
Yes, this is exactly what I want! I am wary of moving away from cucumber as a test runner for BDD scenarios, given the extensive toolchain it offers and lots of things that plug into it, plus it's 'already there'.

I use Serenity-js and can attest to its integration with cucumber. The missing parts are as @jan-molak outlined - there is not integration with the playwright extension for vscode, and that is a shame when you compare the experience of running BDD scenarios via cucumber (even with the excellent Serenity/bdd reports) and running tests via the VS code extension (or - even more apparent maybe - running them via the GREAT new UI mode).

I am totally sold on BDD as a development style, and I love what it brings to the design and exploration process - but I also really like the ease of debugging and the accessibility of tests run via the VS code extension or the UI mode. If these facets could be brought together, I think we would have a really compelling BDD tool suite. I'd be very happy to contribute to this effort too.

**Comment by VitaliyPotapov** [comment]
> Allow me to elaborate. Cucumber is _much_ more than just the Gherkin parser; It has a runner that offers [parallel test execution](https://github.com/cucumber/cucumber-js/blob/main/docs/parallel.md), supports tagging and filtering scenarios using [tag expressions](https://github.com/cucumber/tag-expressions), integrates with [numerous reporters](https://cucumber.io/docs/cucumber/reporting/?lang=javascript#custom-formatters), third-party [plugins](https://cucumber.io/docs/cucumber/reporting/?lang=javascript#third-party-plugins), not to mention project management tools, [code editors](https://behavepro.app/features) and [collaboration tools](https://smartbear.com/blog/introducing-cucumberstudio/). Cucumber.js also has a dedicated [VS Code extension](https://marketplace.visualstudio.com/items?itemName=CucumberOpen.cucumber-official) that supports Gherkin auto-completion.
> 
> Making Playwright Test try to replicate all that seems like a _lot_ of work that could've been avoided by improving the support offered by Playwright to other test tools, such as Cucumber.js.

I think the idea is not to replicate from scratch all things you mentioned. Most of it is already implemented in Playwright runner. It supports [parallel test execution and sharding](https://playwright.dev/docs/test-parallel), [tagging and filtering scenarios](https://playwright.dev/docs/test-annotations#tag-tests), has a lot of [built-in reporters](https://playwright.dev/docs/test-reporters) and [third-party integrations](https://github.com/allure-framework/allure-js/blob/master/packages/allure-playwright/README.md).
All tools for writing BDD scenarios remain the same as there is no need to re-invent Gherkin syntax.

> We can achieve 1. and 3. using BeforeAll and AfterAll hooks... If we wanted to make it more fancy, we could use a custom [World](https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/world.md#real-world-example) and use lazy initialisation to launch the browser and initialise the page

Playwright runner performs all these initializations/cleanup for you automatically - this is one of the main benefits! And with [custom fixtures](https://playwright.dev/docs/test-fixtures#with-fixtures) the code become far more clear and explicit compared to `before*` / `after*` hooks and custom `World` context.

**Comment by NikkTod** [comment]
Guys I understand that you do not want to migrate as you have to set up things again, but playwright runner is already at the level of the cucumber runner if not better. 
You can have your filtering scenarios, integration with custum reporters, parallel test execution, dependency projects, and on top of that very cool features from vs code extention, UI mode and Visual testing. All that is like you install playwright and you have it, no need to integrate or complex set up.
I know that it is difficult to make one tool, which can do anything, but I always prefer not to have so many different frameworks.

For me, with this feauture we just want support for Gehrkin/BDD syntax, we already have all the other things with playwright runner.

I have tried [playwright-bdd](https://github.com/vitalets/playwright-bdd) from and [gherkin-wrapper](https://github.com/Niitch/gherkin-wrapper), both solutions are doing a great job and few things are left to be fixed and implemented.

As far as I understood if playwright team sees that those kind of bdd wrapper projects are actively used they would integrate in future and we would get our in-build support.

**Comment by arvinder06** [comment]
Cucumber js is great with playwright, however we lose the capability to use playwright features like visual testing, sharding etc. Would love to have BDD native to playwright so that we can run BDD tests using playwright runner.

**Comment by devpro** [comment]
[cucumber-tsflow](https://github.com/timjroberts/cucumber-js-tsflow) works really well. I have implemented a solution thanks to `Playwright` (which is so great), `Cubumber.js` and `cucumber-tsflow`.

What is missing is the integration with Playwright CLI (include tests in reports) and Playwright VS Code extension (run/debug).

Here is what it looks like:

- Gherkin: [acceptance/admin-ui/features/login.feature](https://github.com/hobbyfarm/e2e-testing/blob/main/acceptance/admin-ui/features/login.feature)

- Code behind: [acceptance/admin-ui/step-definitions/user-auth.step-definition.ts](https://github.com/hobbyfarm/e2e-testing/blob/main/acceptance/admin-ui/step-definitions/user-auth.step-definition.ts)

It all works fine with Playwright thanks to this two classes:

- [src/execution-flow/base.step-definition.ts](https://github.com/hobbyfarm/e2e-testing/blob/main/src/execution-flow/base.step-definition.ts)

```ts
import { chromium } from '@playwright/test';
import * as dotenv from 'dotenv';
import { after, before, binding } from 'cucumber-tsflow';
import { SessionContext } from './session.context';

@binding([SessionContext])
export class BaseStepDefinition {
  public constructor(protected sessionContext: SessionContext) { }

  @before('@WebApp')
  public async beforeAllWebAppScenarios(): Promise<void> {
    dotenv.config();
    const browser = await chromium.launch();
    this.sessionContext.browser = browser;
    const context = await browser.newContext({ ignoreHTTPSErrors: true });
    this.sessionContext.browserContext = context;
    this.sessionContext.page = await context.newPage();
  }

  @after('@WebApp')
  public async afterAllWebAppScenarios(): Promise<void> {
    await (this.sessionContext.browserContext ?? (() => { throw new Error('browserContext is null'); })()).close();
    await (this.sessionContext.browser ?? (() => { throw new Error('browser is null'); })()).close();
  }
}
```

- [src/execution-flow/session.context.ts](https://github.com/hobbyfarm/e2e-testing/blob/main/src/execution-flow/session.context.ts)

```ts
import { Browser, BrowserContext, Page } from '@playwright/test';

export class SessionContext {
  page: Page | null = null;
  browser: Browser | null = null;
  browserContext: BrowserContext | null = null;
  current: unknown;
}
```

**Comment by NikkTod** [comment]
@devpro great effort on this one, thanks for sharing it, could be very useful for people who want to stay with cucumber js runner. But yet again as you mentioned there are some missing features as with other solutions.

Don`t get me wrong, the cucumber-tsflow is great and I see a lot of people are using it.

The idea with this ticket and the mentioned by me current solutions is not to miss anything from playwright runner features.

**Comment by Tallyb** [comment]
I just looked thru this thread, and was surprised to find out that my repo was not mentioned: https://github.com/Tallyb/cucumber-playwright
It has been around for 2 years and got quite a few forks used as a starter project. (It used to be in the description of the cucumbers channel on Slack, and got lost when moving to Discord). 
While I agree with @jan-molak that cucumber is a full-featured runner, it lags behind PW test runner in more than one aspect. First, you can define a before feature hook (only before scenario). In the same way you can only define background steps for each scenario, but not at the feature level. 
Also, sharding, fixtures, and of course - UI mode.

**Comment by jan-molak** [comment]
@Tallyb - I agree with your assessment, echoed by many others in this thread - there are significant differences between the Cucumber.js runner and Playwright Test. I also share your sentiment towards Playwright Test - I think it offers a brilliant developer experience and I was well impressed seeing how it brought the idea of [PyTest fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) to the world of JavaScript.

It feels like for BDD practitioners, the ideal solution could be:
1. to have Playwright Test _become_ a fully-fledged Cucumber runner, so parse `.feature` files, understand tags, integrate with external tools (like the numerous Jira plugins), provide its excellent DI mechanism in step definitions, and so on. As @VitaliyPotapov mentioned, a lot of the infrastructure needed to achieve this might exist in Playwright Test already. Bringing first-class support for parsing feature files and injecting Playwright fixtures into step definitions would help to bridge the gap and addresses the points brought up by @NikkTod
2. find an alternative that _doesn't_ require folks to use Cucumber or Gherkin at all, yet brings the same benefits of code encapsulation and improved reporting to their Playwright Tests

While I fully support option 1 (just as much as I support improving support for Playwright in the original Cucumber project), I've been working recently on option 2 and wanted to share my results so far in the hope of your feedback.

## When do you need Cucumber/Gherkin?

As opposed to Playwright Test, Cucumber is a [collaboration tool](https://cucumber.io/blog/collaboration/the-worlds-most-misunderstood-collaboration-tool/), not a test automation tool. Its main benefit is that it allows you to express executable specifications of your system in one of the [79 supported languages](https://cucumber.io/docs/gherkin/languages/) instead of just code.

This can be incredibly helpful if your audience prefers to use their native language, like French, Spanish, Arabic, or Japanese, to define their acceptance tests instead of reading your code written in English. Another use case is when your business sponsors/organisation uses external tools like Jira to collaborate on or review Gherkin scenarios without having to access the code or if they simply prefer using Gherkin to express the scenarios instead of doing that directly in code. 

All those cases are perfectly valid and quite popular.

However, what if your intended audience simply needs readable reports and doesn't care how they're generated or what tool you use to make them? What if you use techniques like [Example Mapping](https://cucumber.io/blog/bdd/example-mapping-introduction/) or [Feature Mapping](https://www.manning.com/books/bdd-in-action-second-edition) to discover your requirements and don't actually _need_ to capture them in Gherkin? 

What if you could use Playwright Test to write highly readable code and produce test reports as good as the ones offered by Cucumber? What if it all worked with Playwright Test UI Mode, Trace Viewer, and HTML reports?

## How Serenity/JS Screenplay Pattern improves readability

I've been recently working on integrating [Serenity/JS](https://serenity-js.org) with Playwright Test and leveraging features introduced in Playwright 1.33/1.34 to extend Playwright's reporting capabilities and make its reports more human-friendly.

You can read more about it in my article on [Serenity/JS support for Playwright Test UI Mode](https://serenity-js.org/blog/support-for-playwright-ui-mode/).

In short, I'm leveraging Playwright [`test.step`](https://playwright.dev/docs/api/class-test#test-step) to translate [Serenity/JS Screenplay Pattern](https://serenity-js.org/handbook/design/screenplay-pattern) tasks into Playwright "steps" without the developer ever having to invoke `test.step` directly in their code.

An example todo list app scenario implemented using Serenity/JS Screenplay Pattern and using Playwright Test looks like the listing below (see [full example](https://github.com/serenity-js/serenity-js-playwright-test-template/blob/main/spec/recording-items.spec.ts)):

```typescript
import { describe, it } from '@serenity-js/playwright-test';

describe('Recording items', () => {

    describe('Todo List App', () => {

        it('should allow me to add todo items', async ({ actor }) => {
            await actor.attemptsTo(
                startWithAnEmptyList(),

                recordItem(TODO_ITEMS[0]),

                Ensure.that(itemNames(), equals([
                    TODO_ITEMS[0],
                ])),
            );
        });
    });
});
```

As you can hopefully see above, even though the scenario is not expressed in Gherkin, it's still perfectly readable to an English speaker. Plus, you can run it using the Playwright Test extension for VS Code.

With Serenity/JS, developers compose tasks like `recordItem` from sub-tasks like this:

```typescript
export const recordItem = (name: Answerable<string>): Task =>
    Task.where(d `#actor records an item called ${ name }`,
        Enter.theValue(name).into(newTodoInput()),
        Press.the(Key.Enter).in(newTodoInput()),
        Wait.until(itemNames(), contain(name)),
    )
```

Each low-level sub-task (or an "interaction", in Serenity-speak) maps to a Playwright API call,
which you can see together with developer-defined tasks in Playwright reports (including Playwright UI Mode and Trace Viewer):

<img width="790" alt="serenity-js-3 3 0-playwright-html-report" src="https://github.com/microsoft/playwright/assets/1089173/fe816ac0-9a23-4e9c-8a31-97016995eeef">
<img width="1228" alt="serenity-js-3 3 0-playwright-test-trace-viewer" src="https://github.com/microsoft/playwright/assets/1089173/bf37f2b2-a06a-4812-a60b-b5556a43bb39">
<img width="1392" alt="serenity-js-3 3 0-playwright-ui-mode" src="https://github.com/microsoft/playwright/assets/1089173/7232585e-108c-40c3-b446-515d58c9f580">

If you'd like to explore Serenity/JS approach to writing highly-readable tests, you might want to explore the [Serenity/JS + Playwright Test tutorial](https://serenity-js.org/handbook/web-testing/your-first-web-scenario)

And to reiterate, Serenity/JS Screenplay Pattern APIs for Playwright Test are not a replacement; they're an extension designed to work alongside "native" Playwright scenarios and bring the benefits of highly readable and easy-to-understand test code to Playwright Test suites.

**Comment by Terence625** [comment]
I like cucumber and am using it in my daily work, but I think no need to bind BDD strictly with cucumber and Gerkin syntax. In my opinion, implementing BDD in automation tests is mainly about binding test scenarios and steps in human language with automation test codes, and after test run, generating a test report with test scenarios and steps. Both the input and output of automation test are driven by business requirement, which can be written in Gerkin syntax or any other styles your business analyst/quality analyst like to use.

So after walking through all playwright docs and apis, I found that playwright provided enough flexibility for us to customise the test workflow. Here we can leverage its `test.describe` and `test.step` to achieve BDD implementation, using playwright's todo test example:

[todo-step-definitions.ts](https://github.com/Terence625/playwright-cucumber/blob/main/tests-examples/todo-step-definitions.ts)
```
import { test, expect, type Page } from "@playwright/test";
export const step_createTodoItem = async (page: Page, todo: string) => {
  await test.step(`Create todo: ${todo}`, async () => {
    await page.getByPlaceholder("What needs to be done?").fill(todo);
    await page.getByPlaceholder("What needs to be done?").press("Enter");
  });
};
export const step_checkTodoItem = async (page: Page, todo: string) => {
  await test.step(`Check todo: ${todo}`, async () => {
    await expect(page.getByText(todo)).toBeVisible();
  });
};
export const step_createAndCheckTodoItem = async (page: Page, todo: string) => {
  await step_createTodoItem(page, todo);
  await step_checkTodoItem(page, todo);
};
```
[demo-todo-app.spec.ts](https://github.com/Terence625/playwright-cucumber/blob/main/tests-examples/demo-todo-app.spec.ts)
```
import { test as base, expect, type Page } from "@playwright/test";
import {
  step_createTodoItem,
  step_checkTodoItem,
  step_createAndCheckTodoItem,
} from "./todo-step-definitions";
const test = base.extend({
  page: async ({ page }, use) => {
    await page.goto("https://demo.playwright.dev/todomvc");
    await use(page);
  },
});
test.describe("New Todo", () => {
  test("should allow me to add todo items", async ({ page }) => {
    await step_createTodoItem(page, "buy some cheese");
    await step_checkTodoItem(page, "buy some cheese");
    await step_createTodoItem(page, "feed the cat");
    await step_checkTodoItem(page, "feed the cat");
  });
  test("should allow me to add todo items nested steps", async ({ page }) => {
    await step_createAndCheckTodoItem(page, "buy some cheese");
    await step_createAndCheckTodoItem(page, "feed the cat");
  });
});
```
The above implementation clearly shows test scenarios and steps, with all the benefits from playwright test runner
![image](https://github.com/microsoft/playwright/assets/46805148/fa66f74a-343a-4cf8-ae57-d7770705a667)
![image](https://github.com/microsoft/playwright/assets/46805148/8caa52f2-ec61-47b5-b62c-3606be68687d)

And it can also do something that cucumber cannot do which is nested steps, you can combine multiple small steps into one large step, which is a pain point when using cucumber.

This is just a simple example, there are infinite possibility to customise your own BDD implementation with playwright. I'm sure your can import test scenarios and steps from other test management platform and use source map based approach to translate into playwright titles in `test.describe` and `test.step`

**Comment by jake-davisOVO** [comment]
> Cucumber js is great with playwright, however we lose the capability to use playwright features like visual testing, sharding etc. Would love to have BDD native to playwright so that we can run BDD tests using playwright runner.

Perfectly summed up, are we any closer to having this?

**Comment by jake-davisOVO** [comment]
Thanks for getting back to me @NikkTod, 

I am certainly open to using something like playwright-bdd. I guess my concern is that if the updates for the project suddenly stop, we could be left in a hairy position.  I can see it's being developed by a team of two (not to discredit them at all, it looks amazing) so just creates some risk.

It feels like it would create a lot of reassurance if the Playwright team just baked all of this into the product. 

What do you think?

**Comment by NikkTod** [comment]
@jake-davisOVO, I totally agree, it is better if playwright team start supporting it. As far as I understood they are following closely the development of the bdd projects and could in future integrate one of them within Playwright.

**Comment by Yeya9742** [comment]
Seriously cucumber-js should be supported by the Playwright team, too sad this is taking too long :(

**Comment by RogerDat1** [comment]
playwright test runner should start supporting for BDD framework . We are missing out the potential and benefit of running playwright throuht playwright.config.ts(playwright runner file ) rather we have to use cucumber-js . 
Because of this we are missing out paralle execution , intergrating reporting portal io with our framework . 
using cucumber runner file and playwright as a tool is not making our framework scalable .

Playwright team ... please integrate BDD with playwright and make it native to playwright . so that we can use playwright etst runner to its full potential .

**Comment by dangnguyen92** [comment]
I am really surprised that Playwright does not yet support BDD/Cucumber. It would mean a lot to the testing community if it were integrated.

**Comment by ljelonek** [comment]
I was working on Playwright+Cucumberjs PoC in my project (we are currently using WebDriver.IO but would really love to utilize automated waits and other goodies 🙃). Here is a quick summary of things that prevent us from going live:
- Cubumber does not respect Playwright settings.
  - It is not possible to override assertion timeout globally. I was able to set it up on each assertion (which was nice for a PoC) and possibly, we could write a wrapper library that does that. But this would mean significant code pollution that is not acceptable.
  - It is not possible to override step timeout globally. Your test takes longer than default 30 seconds? Great, your test steps will be skipped and an error will be thrown in AfterScenario (also, you will not have screenshot or anything).
- Logs look awful on test report summary. I guess it is something that I could work on, but colured syntax does not render well initially.

Despite that, I must say that it was a cool ride and I am really looking forward to Playwright BDD or more support for Cucumberjs.

**Comment by jan-molak** [comment]
> * Cubumber does not respect Playwright settings (or the other way around?).

Well, in all fairness to both tools, Cucumber is a separate test runner, so knows nothing about Playwright, nor does Playwright know anything about Cucumber 😄 

>   * It is not possible to override assertion timeout globally. I was able to set it up on each assertion (which was nice for a PoC) and possibly, we could write a wrapper library that does that. But this would mean significant code pollution that is not acceptable.

That's an interesting find. I think it's because Playwright Test creates its [own version of `expect`](https://github.com/microsoft/playwright/blob/f5e6f2e6d234809230a6b09afedf85770d73b8db/packages/playwright-test/src/matchers/expect.ts#L169) so that it can rely on the global [`currentTestInfo`](https://github.com/microsoft/playwright/blob/f5e6f2e6d234809230a6b09afedf85770d73b8db/packages/playwright-test/src/common/globals.ts#L44). Of course, there's no equivalent of that in the Cucumber land, hence the issue.

A Cucumber version of expect would need to rely on Cucumber World or some other global configuration mechanism.

>   * It is not possible to override step timeout globally. 

You can use [`setDefaultTimeout`](https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/timeouts.md) to accomplish that, I believe.

> * Logs look awful on test report summary. I guess it is something that I could work on, but colured syntax does not render well initially.

Yeah, that's because Playwright Test [embeds character escape sequences](https://github.com/microsoft/playwright/blob/f5e6f2e6d234809230a6b09afedf85770d73b8db/packages/playwright-test/src/reporters/markdown.ts#L102) in error diffs so that they look better in VSCode (at least I'm guessing that this was the reason).

It would be nicer for Playwright to attach serialised information about `actual` and `expected` values, a'la Mocha/Jasmine (I reported a similar issue at #26848), but then that would require changes to how the entire error serialisation mechanism works in Playwright. Not to mention it's not trivial to turn actual/expected pair into a nice diff as that depends on the type of assertion, so I fully appreciate the complexity.

You might want to have a look at [Serenity/JS Cucumber + Playwright template](https://github.com/serenity-js/serenity-js-cucumber-playwright-template) if you want [nice reports](https://serenity-js.github.io/serenity-js-cucumber-playwright-template/) and diffs without character escape sequences.

> 
> Despite that, I must say that it was a cool ride and I am really looking forward to Playwright BDD or more support for Cucumberjs.

Agreed. I think to have this high level of integration between Cucumber.js and Playwright like we have with Playwright Test, Playwright would need to create a dedicated runner for Cucumber feature files.

If that were an option, I'd _love_ to see Playwright Test-style dependency injection available in Cucumber steps 🤩

**Comment by JoaoEnesGoncalves** [comment]
it is like something incomplete, for one hand playwright is a great tool but on the other hand it is missing BDD support.
As a tester I see this gap as a major backlog improvement issue.

**Comment by Phonesis** [comment]
Like many here, I've developed a customised framework which uses Playwright to drive Cucumber Feature files. As part of this, Cucumber is the test runner. As a result, we are missing out on all the great Playwright Test runner features. 

A BDD option / config / runner would be absolutely incredible! This is perhaps the only glaring omission Playwright currently has right now. 

Being able to run Cucumber Feature files in parallel using the worker logic which makes Playwright Test so special...

**Comment by med8bra** [comment]
BDD would be a great add-on to Playwright.
Since cucumber has its own test runner, it may be limiting compared to playwright test runner.
It would make sense for playwright to support some sort of BDD testing, or at least offer better integration with cucumber

**Comment by ZhouZhengCarlos** [comment]
Looking forward for this BDD implementation for playwright, in my current job it is useful for our QAs that doesnt have Code expertise

**Comment by ahnpnl** [comment]
This issue is actually a major blocker for my team to switch to Playwright

**Comment by alescinskis** [comment]
> This issue is actually a major blocker for my team to switch to Playwright

I would recommend taking a look at https://github.com/vitalets/playwright-bdd as lack of bdd was a deal breaker for me, but then I've stumbled upon this project.

**Comment by LANDG-MartinP** [comment]
https://github.com/vitalets/playwright-bdd looks interesting but I wonder if there are any limitations? I know, for example, it only recently introduced support for Hooks? 

Does it generate the Playwright HTML report? Can it be used to fully shard tests for parallel execution (test by test, not file by file aka cucumber.js)

**Comment by odeatomas** [comment]
Read the documentation all your questions are covered.

**Comment by jakedavis219** [comment]
I've put together a small framework using playwright and the implementation off bdd/gherkin to organise tests if people want to take a look: https://github.com/jakedavis219/playwright-bdd-typescript-framework

This is the closest I could get without any BDD features being offered by playwright.

**Comment by nokrashy-pwc** [comment]
We need this **ASAP** please!

**Comment by asdfsafd** [comment]
No problem using the playwright test lib with the Cucumber test runner. It's the playwright test runner features we want to benefit from using our Gherkin Scenarios.

**Comment by tcydik** [comment]
> https://github.com/vitalets/playwright-bdd looks interesting but I wonder if there are any limitations? I know, for example, it only recently introduced support for Hooks?
> 
> Does it generate the Playwright HTML report? Can it be used to fully shard tests for parallel execution (test by test, not file by file aka cucumber.js)

This library went a long way since this post. Have a look at it again.

**Comment by MarcHendo** [comment]
Currently using a combination of typescript, playwright and cucumber js. So far its working well, but native support would be great. If it could somehow be integrated into the visual code extension it would take playwright to the next level.

**Comment by acmetro** [comment]
Native support of BDD would be amazing!

**Comment by RichardCariven** [comment]
> Native support of BDD would be amazing!

https://github.com/vitalets/playwright-bdd we are implementing that within our organisation and it is far far far better then any Playwright/Cucumber solution we have tried. 

being able to use Fixtures and the playwright Config opens up so many more opportunities to make our behave tests far more flexible and simpler to maintain and run.

**Comment by Tallyb** [comment]
I was not aware of this: https://pypi.org/project/pytest-bdd/ can be an inspiration.

**Comment by MartinB134** [comment]
But if you want to share steps across multiple folders pytest-bdd will give
you a world of pain. Somehow using a API Setup step for example only works
If you share it via an increasingly bloated conftest.py. Also there is no
ide integration.
It works in the end, but it's painful.

Using behave feels like a massive improvement after switching, for also
multiple other reasons.

Tally Barak ***@***.***> schrieb am Do., 16. Mai 2024, 15:09:

**Comment by LahiruMadhawa2020** [comment]
Eagerly waiting on this feature for a long time. 
Hoping for the best that there is a feasibility to provide it in a early future release.

**Comment by gasparnagy** [comment]
> The last remaining person who was working on the Cucumber project has been laid off:
> 
> https://mattwynne.net/new-beginning

This is misleading. Matt has been laid off _from SmartBear_, but the Cucumber project moves on. All cucumber flavors have regular releases.

**Comment by nilPPanda** [comment]
Is this under development then? Also interested on this. Thanks!

**Comment by anemer** [comment]
Hello,

True question: what do you mean by "_BDD implementation_"?

Do you mean something that make **Playwright Test** runner able to: 
- link steps written in feature files (in _Gherkin_ style) to steps implemented in Playwright (Typescript version?) functions? Like Cucumber.js
- manage a isolated context by scenario (as _World_ notion in Cucumber),
- run those tests/scenarios?

Second question: **why**?

IMHO : 
- it's possible to "do" BDD with any tool. 
- It's already possible to "do" it with Playwright Test (in particular with its notion of `test.step()` that gives clear reports...).
- BDD is not a matter of tools. Because BDD != Gherkin != Cucumber 😉

Those articles explain this  better than me:
- https://www.curiositysoftware.ie/blog/bdd-gherkin-cucumber
- https://specflow.org/bdd/bdd-most-misunderstood-concepts/

**Comment by nilPPanda** [comment]
> Hello,
> 
> True question: what do you mean by "_BDD implementation_"?
> 
> Do you mean something that make **Playwright Test** runner able to:
> 
> * link steps written in feature files (in _Gherkin_ style) to steps implemented in Playwright (Typescript version?) functions? Like Cucumber.js
> * manage a isolated context by scenario (as _World_ notion in Cucumber),
> * run those tests/scenarios?
> 
> Second question: **why**?
> 
> IMHO :
> 
> * it's possible to "do" BDD with any tool.
> * It's already possible to "do" it with Playwright Test (in particular with its notion of `test.step()` that gives clear reports...).
> * BDD is not a matter of tools. Because BDD != Gherkin != Cucumber 😉
> 
> Those articles explain this better than me:
> 
> * https://www.curiositysoftware.ie/blog/bdd-gherkin-cucumber
> * https://specflow.org/bdd/bdd-most-misunderstood-concepts/

I honestly think that when people asking about BDD integration its asking about cucumber step system integration and still use capabilities of playwright runner.

**Comment by asdfsafd** [comment]
We would like to use the playwright test runner with our Cucumber feature files. Using the playwright lib with the Cucumber test runner is working but we would love to be able to use all the playwright tools, too.

**Comment by jan-molak** [comment]
> - it's possible to "do" BDD with any tool.
> - It's already possible to "do" it with Playwright Test (in particular with its notion of test.step() that gives clear reports...).
> - BDD is not a matter of tools. Because BDD != Gherkin != Cucumber 😉

@anemer - I couldn't agree more. BDD is about collaboration and not the tools you use.

> I honestly think that when people is asking about BDD integration, at the end its asking about cucumber step system integration and still use capabilities of playwright runner.

@nilPPanda - I think you're right. From what I'm reading in this thread, it feels like the desire to "do BDD with Playwright" comes down to three needs:
1. Having test reports that explain what happened in the test scenario in a human-readable format, so that they could be shared with the business sponsors akin to Cucumber reports (Playwright reports are great, but they're optimised for developers, not business partners)
2. Having an abstraction on top of the raw Playwright APIs, so that when you read the code it's easy to understand what's going on in [business domain terms](https://blog.mattwynne.net/2013/01/17/the-problem-with-solutions/)
3. Being able to still use the excellent Playwright Test tools and the VSCode / JetBrains IDE plugin

So let's think through that:
1. Plugging Playwright into Cucumber is reasonably easy - you'd to it the same way you'd plug in Selenium or any other web integration tool. I gave examples of how this could be accomplished in my previous comment https://github.com/microsoft/playwright/issues/11975#issuecomment-1567598903
2. Designing good quality abstractions in automated tests can be _really_ challenging, as challenging as designing good quality abstractions in production code. Cucumber step definitions mechanism provides an easy way to introduce basic abstractions that aggregate a bunch of low-level Playwright API calls and give them a nice human-readable description. However, Cucumber step definitions are designed to be simple mappings that don't support code reuse or composition (e.g. you can't call a Cucumber step from another Cucumber step). I completely agree with what @anemer said that developers could use `test.step()` and build their own abstractions around it. However, building test automation frameworks from scratch requires significant experience and time investment. Most teams don't feel like they can afford this as they need to focus on delivering features.
3. This one's tricky, too. Understandably, the VisualStudio Code extension for Playwright, just like the JetBrains plugin for Playwright, is optimised to support Playwright Test runner. Making them support Cucumber would require a significant time and effort investment. 

## What if you really need Cucumber?

So let's go back to @anemer's original question: "do you need Cucumber with Playwright?"

Some teams simply like Cucumber and Gherkin. Some teams use Cucumber because it integrates with their internal reporting, documentation, or auditing systems. Some teams need to write their executable specifications in a [language other than English](https://cucumber.io/docs/gherkin/languages/), and Cucumber offers an easy way to introduce a translation layer.

If this describes your situation, then you'll probably want to stick with Cucumber as a test runner and use Playwright as a web integration tool.

You can, however, introduce Serenity/JS into this mix to improve the reporting, code design, and integration capabilities of your framework and avoid having to build everything from scratch.

Check out this [integration guide](https://serenity-js.org/handbook/test-runners/cucumber/) and this [Serenity/JS Cucumber Playwright project template](https://github.com/serenity-js/serenity-js-cucumber-playwright-template)

## What if you don't need Cucumber but want great reporting and test design?

Now, what if you really want the capabilities and developer experience offered by Playwright Test, great reporting, business-friendly abstractions, but don't want to spend years developing a test automation framework from scratch?

What if I told you that you might not need Cucumber at all, even if your team does BDD 🤔

Have a look at my previous comment at https://github.com/microsoft/playwright/issues/11975#issuecomment-1575231103 explaining an alternative approach of using Serenity/JS with Playwright Test as a business-friendly abstraction layer.

You might also want to check out:
- a [guide on extending existing Playwright Test suites to use Serenity/JS](https://serenity-js.org/handbook/getting-started/serenity-js-with-playwright-test/)
- [Serenity/JS Playwright Test project template](https://github.com/serenity-js/serenity-js-playwright-test-template)

**Comment by gasparnagy** [comment]
I think this feature request is really about to make the Playwright runner extensible in a way that other sub-frameworks can plug in, so that you can use other frameworks (Cucumber, Serenity/JS, etc) to provide test source for the Playwright runner, so that you can use all the benefits that comes form the Playwright runner, but describe your tests in a feature file for example. 

This is the approach that WebdriverIO uses, see https://webdriver.io/docs/frameworks/#using-cucumber.

As many of you have mentioned, you _can_ live and do BDD without this (can use the Cucumber runner, have a BDD-like own naming convention, etc), but still, this would add quite some benefits.

**Comment by jan-molak** [comment]
> I think this feature request is really about to make the Playwright runner extensible in a way that other sub-frameworks can plug in, so that you can use other frameworks (Cucumber, Serenity/JS, etc) to provide test source for the Playwright runner, so that you can use all the benefits that comes form the Playwright runner, but describe your tests in a feature file for example.
> 
> This is the approach that WebdriverIO uses, see https://webdriver.io/docs/frameworks/#using-cucumber.

I agree; it would be amazing if Playwright Test supported a similar mechanism.
There are some interesting challenges here, though. Particularly around supporting parallel execution and making Cucumber compatible with Playwright Test's async dependency injection mechanism (given Cucumber's synchronous nature of the World object -https://github.com/cucumber/cucumber-js/issues/2156).

I think closer integration would be great, and I'd love to collaborate on that; it would, however, require a _lot_ of effort and the maintainers of the various frameworks working together.

**Comment by MartinB134** [comment]
Correct me if I am wrong, but I think we are talking about two different
Things here. There is the playwright Page with it's Tools to interact with
the Browser and there is a test framework with collection of testfunctions,
scoping and testreporting. In terms of Python at least are only talking
about "pytest" the independent unittest framework that is used in the
inspector and the documentation of Playwright. So just changing this to be
used as another option in the inspector and adding a point in the
playwright docu might be enough? You can use nose2, behave, pytest-pdd with
playwright if you "just" import the playwright module and make the page
object available to your test files and test functions.
So If you call your tests with "playwright" you are actually doing a
"pytest" call to collect your test functions from python files which does
not collect .feature files.

Now we want playwright to also "look" for feature files and depending in
that also call either pytest-bdd, unittest or behave instead of pytest or
even implement our own collection mechanism in top of pytest when we do a
"playwright" to run tests.

Am I correct?

Jan Molak ***@***.***> schrieb am Fr., 28. Juni 2024, 12:09:

**Comment by jan-molak** [comment]
> Correct me if I am wrong, but I think we are talking about two different things here.

That's a great point. I think that when saying "Playwright" many people in this thread mean the _"full Playwright experience for Node.js"_, so the combination of all the various Playwright tools for Node.js: VisualStudio Playwright Test Extension, Playwright Test runner, Playwright UI Mode and Playwright web integration library.

Based on my current understanding, [VisualStudio Code Playwright Test Extension](https://github.com/microsoft/playwright-vscode) supports only the Playwright Test runner, and Playwright Test runner is available only for Node.js.

And so, the way I interpret the desire expressed in this thread for Playwright to support Cucumber is that "we'd like Playwright Test Extension and Playwright Test runner to support running and debugging _Cucumber.js_ `.feature` files just as well as it supports running and debugging Playwright Test spec files".

If my interpretation is correct, this feels like a big ask and a significant development effort. For example, VS Code extension would need to integrate with or replicate the behaviour of the [Cucumber extension for VS Code](https://marketplace.visualstudio.com/items?itemName=CucumberOpen.cucumber-official), Playwright Test runner would need to integrate with Cucumber.js, Cucumber.js would need to be made compatible with Playwright's fixtures mechanism, to mention but a few key points.

> You can use nose2, behave, pytest-pdd with playwright if you "just" import the playwright module and make the page object available to your test files and test functions.

Yes, this is what I was referring to in - https://github.com/microsoft/playwright/issues/11975#issuecomment-2195096186
You can use Playwright web integration library for Node.js with Cucumber.js, Mocha, Jasmine, Vitest, or any other JavaScript test runner, the exact same way you'd do it Python - you'd import it just like any other library.

However, not using Playwright Test means you can no longer use the VisualStudio Code Playwright Test Extension, or the [Playwright UI Mode](https://playwright.dev/docs/test-ui-mode).

The approach I recommend to teams using Playwright for Node.js is to use Playwright Test runner, take advantage of Playwright Test extension for VS Code or the Playwright plugin for JetBrains IDEs, and use [Serenity/JS Playwright Test integration](https://serenity-js.org/handbook/getting-started/serenity-js-with-playwright-test/) to introduce business domain language and abstractions and provide more advanced test reporting features.

**Comment by vitalets** [comment]
I'll just add that with `playwright-bdd` you can use Playwright VSCode extension showing spec files that are very close to feature files. Yes, they are not feature files, but you have all tooling you need: run/debug specific scenario, get step with error, view each step execution time, etc.
It was the main reason why we generate Playwright spec files from feature files - to let developers use Playwright tooling as much as possible:

![image](https://github.com/microsoft/playwright/assets/1473072/34725d81-9a1a-492f-a45f-3ab4c6a08403)

**Comment by pramothvm** [comment]
Hi Friends. Do you have the same sample repo project for PlayWright with BDD frame work java. Is there a mail id to reach out to get hep reg this?

**Comment by rishu007** [comment]
this is a must feature please add Gherkin (BDD) support natively in playwright

**Comment by felixpro** [comment]
Good to have BDD with playwright

**Comment by typowy1** [comment]
Good to have cucumber step system integration with playwright.

**Comment by esaari** [comment]
Upvoting this as well.

**Comment by bmulholland** [comment]
Could this be locked? I'd love to get real updates, but endless "upvoting" comments make me want to unsubscribe.

**Comment by otoniel-isidoro** [comment]
For now I would recommend using this https://github.com/vitalets/playwright-bdd. @jan-molak I could make it work to run with serenity/js but the test report is broken...I think I will need to create an adapter to make it work with reporting. I think I will need some explanation about how can I do that.

**Comment by vinodkpasi** [comment]
BDD is must

**Comment by FrancescoBorzi** [comment]
> For now I would recommend using this https://github.com/vitalets/playwright-bdd

Same here. I've been using [Playwright-BDD](https://github.com/vitalets/playwright-bdd) for almost a year already and it works very well.

**Comment by sagarrajput** [comment]
Are there any updates on this, would love to have this feature

**Comment by nathan-hcltech** [comment]
My team also wishes for this feature. We have tried playwright-bdd but the configuration file broke Nx graph. We also want to keep Nx for our monorepos.

**Comment by vitalets** [comment]
@nathan-hcltech could you open an issue in the playwright-bdd repo with more details on your setup and what gets broken?

**Comment by nathan-hcltech** [comment]
> [@nathan-hcltech](https://github.com/nathan-hcltech) could you open an issue in the playwright-bdd repo with more details on your setup and what gets broken?

Thanks for reaching out Vitaliy. When I have the opportunity to test the same setup again, I will create an issue on playwright-bdd repo with details.

**Comment by rishu007** [comment]
Need this feature in my current project please provide support for bdd.

**Comment by rishu007** [comment]
i think Microsoft collected enough feedback for this feature request since last 4 years. Its about taking the decision and implement it

**Comment by pavelfeldman** [comment]
We are not going to be implementing BDD and will prioritize agentic application for test generation and execution.

**Comment by marcindulak** [comment]
@pavelfeldman - BDD and EARS (https://alistairmavin.com/ears/, used by https://kiro.dev/docs/specs/feature-specs/) are probably the main two standardized formats for test specifications used by agents nowadays.

Were you able to evaluate their effectivness when planning for the evolution of the playwright agentic application?
If possible, could you provide more details how it will look like, will it use another format for test specifications, or the human (think business oriented users) readability of tests won't be the focus?

**Comment by PavanMudigondaTR** [comment]
> [@pavelfeldman](https://github.com/pavelfeldman) - BDD and EARS (https://alistairmavin.com/ears/, used by https://kiro.dev/docs/specs/feature-specs/) are probably the main two standardized formats for test specifications used by agents nowadays.
> 
> Were you able to evaluate their effectivness when planning for the evolution of the playwright agentic application? If possible, could you provide more details how it will look like, will it use another format for test specifications, or the human (think business oriented users) readability of tests won't be the focus?

here is there current spec you are looking for

https://github.com/microsoft/playwright/blob/main/examples/todomvc/specs/basic-operations.plan.md

https://playwright.dev/docs/test-agents#introduction
